File: backend/src/main/java/backend/model/services/OrderService.java
package backend.model.services;

import backend.model.entities.ItemBox;
import backend.model.entities.Order;
import backend.model.entities.OrderBox;
import backend.model.exceptions.InstanceNotFoundException;
import backend.model.exceptions.PermissionException;

import java.util.List;

public interface OrderService {

    Order createOrder(Long userId) throws PermissionException, InstanceNotFoundException;

    List<OrderBox> addBoxToOrder(Long userId, Long orderId, Long itemId, int numBoxes, int numItemsInBox)
            throws PermissionException, InstanceNotFoundException;

    Block<Order> findOrderDrafts(Long userId, int page, int size) throws PermissionException, InstanceNotFoundException;

    Order findOrderById(Long userId, Long orderId) throws InstanceNotFoundException, PermissionException;

    List<OrderBox> findOrderBoxesById(Long userId, Long orderId) throws InstanceNotFoundException, PermissionException;

    List<OrderBox> updateNumberOfBoxesInOrder(Long userId, Long orderId, Long orderBoxId, int newNumberOfBoxes)
            throws PermissionException, InstanceNotFoundException;

    List<OrderBox> removeBoxFromOrder(Long userId, Long orderId, Long orderBoxId)
            throws PermissionException, InstanceNotFoundException;

    Boolean deleteOrderById(Long userId, Long orderId) throws PermissionException, InstanceNotFoundException;

    Order sendOrderToAdmins(Long userId, Long orderId) throws PermissionException, InstanceNotFoundException;

    Block<Order> findOrdersSentToAdmins(Long userId, int page, int size) throws PermissionException, InstanceNotFoundException;

    void setOrderDone(Long userId, Long orderId) throws PermissionException, InstanceNotFoundException;

    List<Order> findOrderHistory(Long userid);

    void updateWarehouseStock(Long userId, Long orderId, String warehouseName) throws InstanceNotFoundException, PermissionException;

}

File: frontend/src/backend/orderService.js
import {config, appFetch} from './appFetch';

export const findOrderDrafts = ({page}, onSuccess, onErrors) =>
    appFetch(`/orders/findOrderDrafts?page=${page}`,
        config('GET'), onSuccess, onErrors);

export const findOrderById = (orderId, onSuccess, onErrors) =>
    appFetch(`/orders/${orderId}`,
        config('GET'), onSuccess, onErrors);

export const createOrder = (onSuccess, onErrors) =>
    appFetch(`/orders/createOrder`,
        config('POST'), onSuccess, onErrors);

export const addBoxToOrder = (orderId, itemId, numBoxes, numItemsInBox, onSuccess, onErrors) =>
    appFetch(`/orders/${orderId}/addBox?itemId=${itemId}&numBoxes=${numBoxes}&numItemsInBox=${numItemsInBox}`,
        config(`POST`), onSuccess, onErrors);

export const findBoxesInOrder = (orderId, onSuccess, onErrors) =>
    appFetch(`/orders/${orderId}/boxes`,
        config(`GET`), onSuccess, onErrors);

export const updateNumBoxesInOrder = (orderId, boxId, newNumberOfBoxes, onSuccess, onErrors) =>
    appFetch(`/orders/${orderId}/boxes/${boxId}/updateNumBoxes?newNumberOfBoxes=${newNumberOfBoxes}`,
        config(`PUT`), onSuccess, onErrors);

export const deleteBoxInOrder = (orderId, boxId, onSuccess, onErrors) =>
    appFetch(`/orders/${orderId}/boxes/${boxId}/delete`,
        config(`POST`), onSuccess, onErrors);

export const deleteOrderById = (orderId, onSuccess, onErrors) =>
    appFetch(`/orders/${orderId}/deleteOrder`,
        config('POST'), onSuccess, onErrors);

export const sendOrderToAdmins = (orderId, onSuccess, onErrors) =>
    appFetch(`/orders/${orderId}/sendOrderToAdmins`,
        config('PUT'), onSuccess, onErrors);

export const findOrdersSentToAdmins = ({page}, onSuccess, onErrors) =>
    appFetch(`/orders/findOrdersSentToAdmins?page=${page}`,
        config('GET'), onSuccess, onErrors);

export const updateWarehouseStock = (orderId, warehouseName, onSuccess, onErrors) =>
    appFetch(`/orders/${orderId}/updateWarehouseStock`,
        config('POST', {warehouseName}), onSuccess, onErrors);
